src/util/DeferredProperty.java

```diff
@@ -0,0 +1,231 @@
+     */
+    public DeferredProperty() {
+        this(null, new ArrayList<>());
+    }
+
+    /**
+     * Bypasses the dirty checks and just returns the value of {@link DeferredProperty#value}.
+     *
+     * @return The current value of {@link DeferredProperty#value}.
+     */
+    public T getValue() {
+        return value;
+    }
+
+    /**
+     * Marks this instance as dirty and only if this instance was initially marked as dirty do we propagate this to its {@link DeferredProperty#observers}.
+     *
+     * @return <code>True</code> if we marked this instance as dirty.
+     */
+    public boolean markAsDirty() {
+        // Only when this instance is clean do we want to make it dirty.
+        if (!isDirty()) {
+            isDirty = true;
+
+            // Mark the observers as dirty as they inherit it from this instance.
+            // As it is impossible to add observers after construction then a livelock would never happen.
+            for (DeferredProperty<?> property : observers) {
+                property.markAsDirty();
+            }
+
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Sets the dirty flag as <code>false</code>. If this instance was dirty then true si removed to indicate the dirty flag change from <code>true</code> to <code>false</code>.
+     * Unlike {@link DeferredProperty#markAsDirty()} this does not propagate the cleaning to its observers. For this reason, observers that are dirty will not be cleaned.
+     *
+     * @return <code>True</code> if the dirty flag changed to <code>true</code> from <code>false</code>.
+     */
+    public boolean clean() {
+        if (isDirty()) {
+            isDirty = false;
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * If this instance {@link DeferredProperty#isDirty} then we run the {@link Runnable} and then {@link DeferredProperty#clean()}
+     * This should be useed if the dirty state of this instance can be controled by performing an operation rather than setting a new {@link DeferredProperty#value}.
+     *
+     * @param runnable The runnable which is run if we cleaned the object.
+     * @return <code>True</code> if the dirty flag changed to <code>true</code> from <code>false</code>.
+     */
+    public boolean clean(Runnable runnable) {
+        boolean cleaned = clean();
+        if (cleaned) {
+            runnable.run();
+        }
+        return cleaned;
+    }
+
+    /**
+     * Used to check if this instance is marked as dirty.
+     *
+     * @return True if this instance is dirty.
+     */
+    public boolean isDirty() {
+        return isDirty;
+    }
+
+    /**
+     * Used to check if this instance is _not_ marked as dirty.
+     *
+     * @return True if this instance is not dirty.
+     */
+    public boolean isClean() {
+        return !isDirty();
+    }
+
+    /**
+     * Set the {@link DeferredProperty#value} and cleans this instance by not marking it as dirty.
```

Comment:
```suggestion
     * Sets the {@link DeferredProperty#value} and cleans this instance by marking it as not dirty.
```

You use two verb tenses in the same sentence (`set` vs `cleans`). Also, notice the difference between `not marking it as dirty` and `marking it as not dirty`.

You could also consider removing `by not marking it as dirty`, as how cleaning is performed is not really relevant here.